Shop_Unity_base/Assets/Scripts/Shop/Model/Item.cs

```diff
@@ -1,38 +1,80 @@
-﻿using Model.Items;
-
+﻿
 namespace Model
 {
-    // This class holds data for an Item. Currently it has a name, an iconName and an amount.
-    public abstract class Item : IClonable, ITradeable
+    using System;
+    using Model.Items;
+    using System.Collections.Generic;
+
+    public abstract class Item : IClonable, ITradeable, IComponentOwner<Item>
     {
         public enum ItemType { Potion, Weapon }
 
+        protected Dictionary<Type, Component<Item>> TypeToComponent = new Dictionary<Type, Component<Item>>();
+
         protected string name;
-        protected string iconName;
         protected int cost;
 
+        public string Name => name;
+        public int Cost => cost;
+
         protected Item()
         {
-            iconName = "item";
         }
 
         protected Item(string name, int cost)
         {
             this.name = name;
             this.cost = cost;
-            iconName = "item";
         }
 
-        protected virtual void CloneFields(Item item)
+        public void AddComponent(Component<Item> component)
+        {
+            if (TypeToComponent.ContainsKey(component.GetType()))
+                return;
+
+            Type type = component.GetType();
+            TypeToComponent.Add(type, component);
+            component.SetOwner(this);
+        }
+
+        public void AddComponents(params Component<Item>[] components)
+        {
+            for (int i = 0; i < components.Length; i++)
+                AddComponent(components[i]);
+        }
```

Comment:
Your Get and Remove methods use generics, why not here? I do realize that this might be done to avoid having Component Owners construct their own components, but I don't see a problem if they do? You might elaborate to me or change it?

Reply:
The current implementation needs the user to construct components with specific parameters to create different objects. If the code were more of a black box (which Factory is already helping with) this would be a good change, but I want to allow someone to open Factory and add more custom Item prototypes.